tests/actions/test_migrate.py

```diff
@@ -453,68 +457,70 @@ def _mkstemp(*args, **kwargs):
+            dump_yaml(pipeline_file, doc, style)
             return subprocess.CompletedProcess(cmd, 0, stdout="", stderr="")
 
         monkeypatch.setattr("tempfile.mkstemp", _mkstemp)
         monkeypatch.setattr("subprocess.run", subprocess_run)
 
-        if chdir:
-            monkeypatch.chdir(tmp_path)
-            manager.apply_migrations()
+        monkeypatch.chdir(tmp_path)
+        op = MigrationFileOperation(self.package_file.task_bundle_upgrades)
+        op.handle(self.package_file.file_path)
 
-            expected = ["echo add a new task", "echo hello", "echo remove task param"]
-            assert test_context["executed_scripts"] == expected
+        doc = load_yaml(self.package_file.file_path)
+        if "kind: PipelineRun" in pipeline_and_run_yaml:
+            tasks = doc["spec"]["pipelineSpec"]["tasks"]
         else:
-            with pytest.raises(ValueError, match="Pipeline file does not exist: .+"):
-                manager.apply_migrations()
+            tasks = doc["spec"]["tasks"]
 
-    def test_raise_error_if_migration_process_fails(
-        self, pipeline_yaml, caplog, monkeypatch, tmp_path
-    ):
-        caplog.set_level(logging.DEBUG, logger="migrate")
+        assert tasks[-1] == {"name": "test"}
 
-        renovate_upgrades = [
-            {
-                "depName": TASK_BUNDLE_CLONE,
-                "currentValue": "0.1",
-                "currentDigest": "sha256:cff6b68a194a",
-                "newValue": "0.2",
-                "newDigest": "sha256:96e797480ac5",
-                "depTypes": ["tekton-bundle"],
-                "packageFile": ".tekton/pipeline.yaml",
-                "parentDir": ".tekton",
-            },
-        ]
-        manager = TaskBundleUpgradesManager(renovate_upgrades, SimpleIterationResolver)
+        # Verify the original formatting is preserved
+        assert pipeline_and_run_yaml in Path(self.package_file.file_path).read_text()
 
-        # Not really resolve migrations. Mock them instead, then apply.
-        tb_upgrade = list(manager._task_bundle_upgrades.values())[0]
-        c = Container(tb_upgrade.dep_name)
+    def test_do_not_save_if_no_changes(self, pipeline_and_run_yaml, monkeypatch, tmp_path):
+        self.prepare(tmp_path, pipeline_and_run_yaml)
 
-        c.tag = tb_upgrade.new_value
-        c.digest = generate_digest()
-        tb_upgrade.migrations.append(
-            TaskBundleMigration(task_bundle=c.uri_with_tag, migration_script="echo add a new task")
-        )
+        counter = itertools.count()
 
-        c.tag = tb_upgrade.new_value
-        c.digest = tb_upgrade.new_digest
-        tb_upgrade.migrations.append(
-            TaskBundleMigration(
-                task_bundle=c.uri_with_tag, migration_script="echo remove task param"
-            )
+        expected_dump_yaml_calls = 0  # for Pipeline
+        if "kind: PipelineRun" in pipeline_and_run_yaml:
+            expected_dump_yaml_calls = 1
+
```

Comment:
Why do we only expect changes for PipelineRuns?

Reply:
Ah, I see. That's quite confusing, is there any way we could make the test clearer?

If not, at least a comment would help